### ISSUE TITLE
Design/execution: Dataset v3 for diversity-preserving incremental Lean training

### ISSUE THREAD

**Issue by murillo128** [murillo128/qwen-lean#82]
## Goal

Build a new canonical **Dataset v3** for the next qwen-lean executor training run. This issue owns the **data model, extraction, serialization, derived training views, split isolation, verification, and corpus diagnostics only**. Model training belongs in a separate dependent issue.

The main design change is to stop treating `theorem -> one normalized whole proof` as the only optimizer-visible task. Dataset v3 must preserve original proof diversity and add **incremental proof-continuation examples** so the executor learns to continue a valid proof trajectory rather than collapsing onto one frequent whole-proof template.

## Why v3 is needed

Issue #78 / PR #79 produced useful engineering infrastructure but exposed a model failure mode that should change the corpus design:

- Base solved 6/415 on `fresh-composition-test-v2`; Q4 solved 0/415.
- The Q4-specific HF/PEFT ↔ vLLM canary passed, so the collapse is not explained by an adapter-load bug.
- Q4 final failures were overwhelmingly short EOS-terminated outputs and almost all used the same `exact ⟨…⟩` constructor mode.
- A post-hoc Q1 trajectory run also reached 0/415 and 0/3320 Lean-valid candidates with 3320/3320 EOS completions, placing the narrowing already in the first quarter of SFT rather than only at Q4.
- Dataset v2 has almost one proof variant per statement and mechanically rewrites a very large fraction of source proofs into `...-to-exact` forms. That representation is formally valid but is now a plausible source of stylistic mode collapse.

Treat v2 as historical evidence. Do not mutate its files or official results.

### 1. Preserve source proof form by default

Re-extract optimizer-visible proofs from their pinned sources rather than using Dataset-v2 transformed completions as the training target when the original proof text is available.

- Preserve tactic blocks, `calc`, `constructor`, `refine`, `apply`, `rw`, `simp`, `induction`, term proofs, branches, etc.
- Do not mechanically canonicalize a term/tactic proof into `exact ...` merely to obtain a uniform surface form.
- Allow a transformation only when it is technically necessary to produce a valid standalone training record; record the transformation explicitly and keep transformation-specific mass measurable.
- Add corpus diagnostics for first proof construct / major proof-form families so a single surface form cannot silently dominate the optimizer mass.

### 2. Whole-proof + incremental proof continuation

For a source proof with structural segments `A B C D`, support both:

```text
theorem                  -> A B C D
theorem + A              -> B C D
theorem + A B            -> C D
theorem + A B C          -> D
```

This is **proof-prefix continuation**, not arbitrary token masking.

Requirements:

- split only at Lean-aware structural boundaries;
- prefer boundaries after complete top-level tactics, completed `have`/`calc` steps, branch boundaries, or other parser/syntax-aware units;
- do not cut inside tokens, syntax atoms, or an unfinished tactic command;
- for proofs where robust structural segmentation is unavailable, retain the whole-proof example rather than inventing unsafe cuts;
- concatenating the stored prefix and target continuation must reconstruct a Lean-valid proof in the same theorem/environment;
- every materialized incremental example must be mechanically attributable to its source whole proof and boundary.

The first v3 implementation does **not** need to expose raw Lean goal states as model input. Prefix continuation is the initial interface; explicit proof-state conditioning can be a later extension.

### 3. Statement-normalized training mass

Incrementalization must not accidentally make long proofs dominate training.

For each theorem, define a bounded total optimizer mass independent of:

- proof length;
- number of valid cut points;
- number of retained source proof variants.

Distribute that theorem-level mass across whole-proof and incremental examples deterministically. Preserve the current useful principle that adding more variants/cut points does not multiply the theorem's aggregate weight.

The exact whole-proof vs continuation mass split should be declared in the Dataset-v3 manifest and be easy to change in the later training issue without rebuilding source extraction where possible. Do not silently choose a ratio because one proof happened to have many boundaries.

### 4. Retain genuinely different valid proofs when available

The schema must support multiple verified proofs per theorem without treating cosmetic rewrites as meaningful diversity.

- Retain naturally occurring independent proof variants where source provenance supports them.
- If additional alternatives are mined later, admit them only after Lean verification.
- Compute a structural/normalized proof fingerprint so whitespace, parentheses, declaration-name aliases, or other cosmetic changes do not masquerade as proof diversity.
- Model-generated alternate-proof mining is optional for the first Dataset-v3 build; do not make a large inference campaign a prerequisite for delivering the core corpus.

### 5. New clean validation/test material

The v2 `fresh-composition-test-v2` has now been inspected and used for post-hoc trajectory diagnosis. It must not become the official selection/test set for v3.

Dataset v3 must provide newly frozen evaluation material with strict isolation:

- no optimizer-visible statement overlap;
- no proof overlap;
- no derivation-family/fingerprint leakage across train/validation/test for synthetic composition families;
- newly generated composition families/instances rather than renamings of v2 held-outs;
- include direct / branching / deep composition and logical shapes such as `iff` that exposed weaknesses in v2;
- Lean-verify all target proofs before freezing membership;
- keep the final v3 test sealed from checkpoint selection and hyperparameter decisions.

Existing miniF2F and v2 fresh results may remain **historical/diagnostic comparators**, but they are not the new clean v3 selection oracle.

### 6. Corpus diagnostics that directly guard against v2's failure mode

Publish aggregate diagnostics with denominators for both raw records and **optimizer mass**:

- whole-proof vs incremental mass;
- original-source vs transformed proof mass;
- first proof construct / proof-form histogram (`exact`, `constructor`, `refine`, `apply`, `intro`, `rw`, `simp`, `calc`, `induction`, etc.);
- proof-length and continuation-length distributions;
- number of continuation boundaries per theorem;
- number of proof variants per theorem and structurally unique proof variants;
- synthetic/real source mass and composition-class balance;
- theorem-level effective weights;
- maximum/minimum examples and mass contributed by any theorem;
- duplicate/cosmetic-proof rates;
- train/validation/test leakage audits.

No corpus is acceptable if a transformation or derived-example policy makes one superficial proof form dominate without being explicitly visible in these diagnostics.

## Reproducibility / package contract

Produce a local canonical package analogous to Dataset v2, with committed manifests/config/evidence sufficient to reproduce it. Large generated corpus files may remain local if that is the repository convention, but their exact hashes, source revisions, generation parameters, counts, and rebuild procedure must be committed.

Pin the same Lean/mathlib/source revisions unless a deliberate environment update is separately justified. No silent truncation or dropped long-context examples.

The package should distinguish:

- canonical source theorem/proof records;
- proof variants;
- structural continuation boundaries;
- derived optimizer examples/views;
- validation/test memberships;
- diagnostics/verification evidence.

Avoid storing redundant full copies of the same source proof for every continuation when a reference + boundary representation can materialize the training row deterministically.

## Verification gates

Before Dataset v3 can unblock training:

1. every canonical whole proof verifies in its pinned Lean environment;
2. every incremental example reconstructs a verifying whole proof from `prefix + continuation`;
3. all derived examples have deterministic identities and source provenance;
4. statement-normalized mass invariants pass exactly;
5. no train/validation/test statement/proof/derivation-family leakage exists under the declared policy;
6. no placeholder (`sorry`/`admit`) reaches optimizer membership;
7. no silent truncation/drop occurs;
8. proof-style / transformation / incremental-mass diagnostics are published;
9. a small materialization fixture demonstrates whole-proof and incremental examples end-to-end.

## Deliverables

- Dataset-v3 schema/config and build pipeline;
- source-preserving proof extraction path;
- Lean-aware incremental-boundary extraction/materialization;
- statement-normalized weighting/view generation;
- new clean v3 validation/test composition memberships;
- verification + leakage checks;
- corpus diagnostics focused on proof-style diversity and mode-collapse risk;
- canonical manifest/evidence and concise README/design notes;
- tests for the invariants above.

## Explicit non-goals

- no Qwen training in this issue;
- no checkpoint selection;
- no Mathia/planner training;
- no RL/GRPO/RLVR;
- do not add v2 final-test failures themselves to training;
- do not optimize the corpus by looking at individual sealed v3 test outcomes.

## Completion criterion

Dataset v3 is complete when a later training job can consume a deterministic, verified mix of source-preserving whole proofs and structurally valid incremental continuations with theorem-normalized mass, while the repository can prove that the v2-style `exact` surface-form collapse is no longer being introduced silently by corpus serialization.
